services/QuillLMS/client/app/bundles/Diagnostic/utils/saveSession.tsx

```diff
@@ -0,0 +1,83 @@
+import * as React from 'react'
+import Pusher from 'pusher-js';
+
+import { getConceptResultsForAllQuestions } from '../libs/conceptResults/diagnostic';
+import { requestPost, requestPut, } from '../../../modules/request/index';
+import { roundValuesToSeconds, } from '../../Shared/index';
+
+const TITLE_CARD_TYPE = "TL"
+
+export const saveSession = (sessionID, timeTracking, playDiagnostic, match, isTurkSession, updateState) => {
+  const { params } = match
+  const { diagnosticID } = params
+
+  updateState({ error: false, });
+
+  const relevantAnsweredQuestions = playDiagnostic.answeredQuestions.filter(q => q.questionType !== TITLE_CARD_TYPE)
+  const results = getConceptResultsForAllQuestions(relevantAnsweredQuestions);
+  const data = { time_tracking: roundValuesToSeconds(timeTracking), }
+
+  if (sessionID) {
+    finishActivitySession(sessionID, results, 1, data, isTurkSession, updateState);
+  } else {
+    createAnonActivitySession(diagnosticID, results, 1, data, isTurkSession, updateState);
+  }
+}
+
+const initializeSubscription = (activitySessionUid, isTurkSession, updateState) => {
+  if (process.env.NODE_ENV === 'development') {
+    Pusher.logToConsole = true;
+  }
+  if (!window.pusher) {
+    window.pusher = new Pusher(process.env.PUSHER_KEY, { cluster: process.env.PUSHER_CLUSTER });
+  }
+  const channel = window.pusher.subscribe(activitySessionUid);
+
+  channel.bind('concept-results-saved', () => {
+    if (!isTurkSession) { document.location.href = `${process.env.DEFAULT_URL}/activity_sessions/${activitySessionUid}` }
+    updateState({ saved: true, });
+  });
+
+  channel.bind('concept-results-partially-saved', () => {
+    if (!isTurkSession) { document.location.href = process.env.DEFAULT_URL; }
+  });
+}
+
+const finishActivitySession = (sessionID, results, score, data, isTurkSession, updateState) => {
+  initializeSubscription(sessionID, isTurkSession, updateState)
+
+  requestPut(
+    `${process.env.DEFAULT_URL}/api/v1/activity_sessions/${sessionID}`,
+    {
+      state: 'finished',
+      concept_results: results,
+      percentage: score,
+      data
+    },
+    (body) => {
+      // doing nothing here because the Pusher subscription should handle a redirect once concept results are saved
+    },
+    (body) => {
+      updateState({
+        saved: false,
+        error: body.meta.message,
+      });
+    }
+  )
+}
+
+const createAnonActivitySession = (lessonID, results, score, data, isTurkSession, updateState) => {
+  requestPost(
+    `${process.env.DEFAULT_URL}/api/v1/activity_sessions/`,
+    {
+      state: 'finished',
+      activity_uid: lessonID,
+      concept_results: results,
+      percentage: score,
+      data
+    },
+    (body) => {
+      initializeSubscription(body.activity_session.uid, isTurkSession, updateState)
+    }
```

Comment:
I don't know of a better way to do this, but I can imagine some weird edge case where this would result in a race condition.  If the controller from this POST call kicks off a job that somehow finishes and triggers Pusher before the controller returns the UID, the Pusher even will be missed because the subscription won't be in place yet.

Again, not sure there's a better way to do this, but I did want to flag this possibility in case someone thought it was something more than just an unlikely edge case that we don't need to worry about.

Reply:
Yeah, this occurred to me too. This is the same pattern we follow with Connect/Proofreader/Grammar, and as far as I know it hasn't been a problem there, but it is possible in theory. I'm open to changing it if someone has an alternate approach, but I couldn't think of anything since we do need the session uid in order to key the subscription off of.

Reply:
Since we're already following the same pattern with connect/proofreader/grammar, I'd probably be inclined to keep as-is for now.

If you wanted to guard from the aforementioned race conditions, I think you could create the uid client side first, then initialize pusher with the value, and finally create the activity session with the uid:

```
const createAnonActivitySession = (lessonID, results, score, data, isTurkSession, updateState) => {
  const activitySessionUid = createUID(). // either use external library or make some method
  initializeSubscription(activitySessionUid, isTurkSession, updateState);

  requestPost(
    `${process.env.DEFAULT_URL}/api/v1/activity_sessions/`,
    {
      activity_uid: lessonID,
      concept_results: results,
      percentage: score,
      data,
      activity_session_uid: activitySessionUid  // Include the UUID in the POST data
    },
```

Reply:
Yeah, this is a good point. I'll make a backlog task to pull this logic out further at some point so we can use it across all of the apps, and then maybe it makes sense to add that functionality then so we introduce it everywhere at once.